react/features/video-quality/subscriber.js

```diff
@@ -193,43 +193,114 @@ function _updateReceiverVideoConstraints({ getState }) {
     const maxFrameHeight = Math.min(maxReceiverVideoQuality, preferredVideoQuality);
     const { remoteScreenShares } = state['features/video-layout'];
     const { visibleRemoteParticipants } = state['features/filmstrip'];
+    const { flags: { sourceNameSignaling } } = state['features/base/config'];
+    const tracks = state['features/base/tracks'];
 
     const receiverConstraints = {
         constraints: {},
         defaultConstraints: { 'maxHeight': VIDEO_QUALITY_LEVELS.NONE },
         lastN,
-        onStageEndpoints: [],
-        selectedEndpoints: []
+        ...sourceNameSignaling ? { prioritizedSources: [] } : { onStageEndpoints: [] },
+        ...sourceNameSignaling ? { selectedSources: [] } : { selectedEndpoints: [] }
     };
 
-    // Tile view.
-    if (shouldDisplayTileView(state)) {
-        if (!visibleRemoteParticipants?.size) {
-            return;
+    if (sourceNameSignaling) {
+        const participantIdToTrackSourceName = tracks.reduce((acc, { jitsiTrack, participantId }) => {
+            const local = jitsiTrack.isLocal();
+            const isVideo = jitsiTrack && jitsiTrack.type === 'video';
+
+            if (!local && isVideo) {
+                const sourceName = jitsiTrack.getSourceName && jitsiTrack.getSourceName();
+
+                if (sourceName) {
+                    acc[participantId] = sourceName;
+                }
+            }
+
+            return acc;
+        }, {});
+
+        let visibleRemoteTrackSourceNames;
+
+        if (visibleRemoteParticipants?.size) {
+            visibleRemoteTrackSourceNames = [ ...visibleRemoteParticipants ].reduce((acc, participantId) => {
+                if (participantIdToTrackSourceName[participantId]) {
+                    acc.push(participantIdToTrackSourceName[participantId]);
```

Comment:
I think you should be able to call [this selector] here instead of creating the mapping above or am I missing something?

[this selector]: https://github.com/jitsi/jitsi-meet/blob/162a67fe8b55c675abb9335113fcd9556720420d/react/features/base/tracks/functions.js#L393

Did you mean to create and store the mapping as an optimization?

Reply:
Thank you, I didn't know we had this selector available to us! So for each participant id, do we need to pull all the tracks with the `presenter` and `video` media type? 

Yes, I was wondering where we can store this mapping and if it is even necessary. We can just use the selector you've mentioned or a slightly modified version. Also, I currently have each participant mapped to only one video track where instead I think it should be an array of tracks. (e.g. camera and screenshare)

Reply:
Yes, you're thinking correctly that in the final solution we'll be using camera and screenshare, but we'll add it together with the code for the new presenter mode. For now let's just pick the `video` tracks (at this point we don't care if it's screen or camera, because there's just one).

I was looking into media types `video` and `presenter` and I think `presenter` is only used locally, so just select for the `video` media type. Hopefully we'll be able to get rid of the presenter type with the new implementation.

Reply:
Got it, I'll make those changes